File: src/magyarlanc/gui.clj
(ns magyarlanc.gui
    (:import [java.awt BorderLayout Dimension Insets Toolkit]
             [java.awt.event ActionListener]
             [java.awt.image BufferedImage]
             [javax.swing BoxLayout ImageIcon JButton JFrame JLabel JPanel JTextArea JTextField]
             [javax.swing.border EmptyBorder])
    (:import [com.googlecode.whatswrong SingleSentenceRenderer]
             [com.googlecode.whatswrong.io CoNLL2009])
    (:import [magyarlanc Dependency HunSplitter])
    (:gen-class))

(defn- conll [lines]
    (map #(let [[a b c _ e _ g h] % _ "_"] (list a b c e _ _ _ _ g _ h _ _ _)) lines))

(defn- whats-wrong [lines]
    (let [renderer (SingleSentenceRenderer.)
          instance (.create (CoNLL2009.) (conll lines))]
        (let [image (BufferedImage. 1 1 BufferedImage/TYPE_4BYTE_ABGR)
              dim (.render renderer instance (.createGraphics image))

              image (BufferedImage. (+ (.getWidth dim) 5) (.getHeight dim) BufferedImage/TYPE_4BYTE_ABGR)
              dim (.render renderer instance (.createGraphics image))]
            image)))

(defn- pretty [lines]
    (apply str (mapcat conj (map #(vec (interpose \tab %)) lines) (repeat \newline))))

(defn- centered [component]
    (let [v (.. Toolkit getDefaultToolkit getScreenSize)
          w (.getPreferredSize component)
          x (/ (- (.getWidth v) (.getWidth w)) 2)
          y (/ (- (.getHeight v) (.getHeight w)) 2)]
        (doto component (.setLocation x y))))

(defn init [sentence]
    (let [frame  (JFrame. "magyarlanc 2.0")
          panel  (JPanel.)
          input  (JTextField. sentence)
          button (JButton. "OK")
          label  (JLabel.)
          output (JTextArea.)]

        (let [v (.. Toolkit getDefaultToolkit getScreenSize)]
            (doto frame (.setDefaultCloseOperation JFrame/DISPOSE_ON_CLOSE) (.setLayout (BorderLayout.))
                        (.setPreferredSize (Dimension. (- (.getWidth v) 150) (- (.getHeight v) 150))) (.setResizable false)))

        (doto panel (.setBorder (EmptyBorder. 15 15 15 15)) (.setLayout (BoxLayout. panel BoxLayout/X_AXIS)))
        (doto label (.setHorizontalAlignment JLabel/CENTER) (.setVisible false))
        (doto output (.setMargin (Insets. 10 10 10 10)) (.setVisible false))

        (doto frame (.add (doto panel (.add input) (.add button)) "North") (.add label "Center") (.add output "South"))

        (.addActionListener button (reify ActionListener (actionPerformed [_ actionEvent]
            (let [sentence (.getText input)]
                (if-not (empty? sentence)
                    (let [lines (Dependency/depParseSentence (first (HunSplitter/splitToArray sentence)))]

                        (doto label (.setIcon (ImageIcon. (whats-wrong lines))) (.setVisible true))
                        (doto output (.setText (pretty lines)) (.setVisible true))

                        (doto (centered frame) .pack (.setVisible true))))))))

        (doto (centered frame) .pack (.setVisible true))))

(defn -main []
    (init "Nehéz lesz megszokni a sok üres épületet, de a kínai áruházak hamar pezsgővé változtathatják a szellemházakat."))

File: src/magyarlanc/GUI.java
package magyarlanc;

import java.awt.BorderLayout;
import java.awt.Component;
import java.awt.Dimension;
import java.awt.Insets;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.IOException;

import javax.imageio.ImageIO;
import javax.swing.BoxLayout;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JTextArea;
import javax.swing.JTextField;
import javax.swing.border.EmptyBorder;

public class GUI
{
    private static JTextArea textarea;
    private static JLabel imageLabel;

    private static String _sentenceAsString(String[][] array)
    {
        StringBuilder sb = new StringBuilder();

        for (int i = 0; i < array.length; i++)
        {
            for (int j = 0; j < array[i].length; j++)
            {
                sb.append(array[i][j]).append('\t');
            }
            sb.append('\n');
        }

        return sb.toString();
    }

    private static void _moveToCenter(Component component)
    {
        component.setLocation(
            (int) ((Toolkit.getDefaultToolkit().getScreenSize().getWidth() - component.getPreferredSize().getWidth()) / 2),
            (int) ((Toolkit.getDefaultToolkit().getScreenSize().getHeight() - component.getPreferredSize().getHeight()) / 2));
    }

    public static void init()
    {
        JFrame frame = new JFrame("magyarlanc 2.0");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.getContentPane().setLayout(new BorderLayout());

        JPanel inputPanel = new JPanel();
        inputPanel.setLayout(new BoxLayout(inputPanel, BoxLayout.X_AXIS));
        inputPanel.setBorder(new EmptyBorder(15, 15, 15, 15));

        JTextField inputField = new JTextField(
            "Nehéz lesz megszokni a sok üres épületet, de a kínai áruházak hamar pezsgővé változtathatják a szellemházakat.");

        JButton sendButton = new JButton("OK");
        sendButton.addActionListener(new ActionListener()
        {
            public void actionPerformed(ActionEvent actionEvent)
            {
                if (inputField.getText() != null && !inputField.getText().equals(""))
                {
                    String[] sentence = HunSplitter.splitToArray(inputField.getText())[0];

                    String[][] depParsed = Dependency.depParseSentence(sentence);

                    BufferedImage bufferedImage = null;
                    try
                    {
                        bufferedImage = ImageIO.read(new ByteArrayInputStream(WhatsWrong.exportToByteArray(depParsed)));
                    }
                    catch (IOException e)
                    {
                        e.printStackTrace();
                    }

                    if (imageLabel != null)
                        imageLabel.setVisible(false);

                    imageLabel = new JLabel(new ImageIcon(bufferedImage));
                    frame.getContentPane().add(imageLabel, "Center");

                    if (textarea != null)
                        textarea.setVisible(false);

                    textarea = new JTextArea();
                    textarea.setText(_sentenceAsString(depParsed));
                    textarea.setMargin(new Insets(10, 10, 10, 10));
                    frame.getContentPane().add(textarea, "South");

                    _moveToCenter(frame);
                    frame.pack();
                    frame.setVisible(true);
                }
            }
        });

        inputPanel.add(inputField);
        inputPanel.add(sendButton);

        frame.getContentPane().add(inputPanel, "North");
        Dimension dim = Toolkit.getDefaultToolkit().getScreenSize();

        frame.setPreferredSize(new Dimension((int) dim.getWidth() - 150, (int) dim.getHeight() - 150));
        frame.setResizable(false);

        _moveToCenter(frame);

        frame.pack();
        frame.setVisible(true);
    }

    public static void main(String[] args)
    {
        init();
    }
}
